### DISCUSSION TITLE
Any interest in CMake + Conan support?

### DISCUSSION THREAD

**Discussion by Twon** [rui314/mold#429] [Ideas]
Hello,

I've been building Mold on Mac & Centos systems so have had to find a solution for the project dependencies across different environments. As such I've created a branch that fetched dependencies from the Conan package manager & builds in CMake (and thus in any of the supported underlying build systems). 

However support is currently partial, it compiles and builds but currently lacks CTest support for all of the test scripts and an install step for packaging (please feel free to try here: https://github.com/Twon/mold/tree/cmake_and_conan).  Do you have any interest in these changes upstream?  If so I'm happy to commit to completing a full conversion if this would be useful (i.e for example I see there is an issue for Windows support, this could be useful as a prerequisite to building under Visual Studio / VSCode).

Let me know if you have any questions and I would be happy to hear your thoughts on this.

Cheers

**Comment by rui314** [comment]
Sorry for the late reply.

I want to migrate to CMake eventually but at the same time I want to keep the build dependencies as small as possible. Currently, we use the good old Makefile, and that's still a lingua franca among Unix programmers. As long as we are working on Unix and macOS, I believe we can live with that. When we start working on Windows support, we need to migrate to CMake, but it feels that that's too early to make a change to this place.

**Comment by Twon** [comment]
Understood, for now I'll maintain a fork of Mold with CMake + Conan.  Should you want any assistance when you get to this task then please feel free to assign the ticket to me in github and I'm happy to pick up any required work.

**Comment by BenBE** [comment]
I think CMake itself for the build is fine. I just don't see the need to use Conan as the package manager when every decent Linux distribution comes with a package manager handling build dependencies for you (you tell it what you need and the package manager puts the files where cmake can find them). I'm always vary of additional tools just to do the job that my system already can do properly.

As for Windows support: CMake works nicely on Windows (usually using it together with Ninja, but I heard the MSVC backend should do just fine too).

**Comment by Twon** [comment]
I actually put together this branch due to difficulties encountered with using the inbuilt package manager under Centos to install the required dependencies.  Perhaps I'm missing something here but it appeared that different instructions for installing packages are required for each variant of Linux and the reliability/availability of packages is dependent on the package manager in question?

In my personal use-case (building across MacOS, Ubunto & Centos) Conan provided a simple solution for consistent cross-platform 
(and cross Linux distribution) behaviour.  As things stand package management for Windows is non-standard and will require the installation of a separate package management system like Chocolatey or vcpkg.

**Reply by BenBE** [reply]
Haven't had any issues with dependencies on Linux (mostly working on Debian+Ubuntu) yet.

But if you found that Conan might help with installing missing dependencies I'm fine with this as long as I'm not forced to use it when I don't need it.

AFAIK there's always an option with cmake to just specify hints for where it should look for packages. Should work on Windows too (even if it is a PITA).

**Reply by Twon** [reply]
It is possible to use Conan so as to not force its usage on the project. This is achieved by using the cmake_find_package generator in Conan as follows: https://github.com/Twon/mold/blob/1d4fce72c6e70a6ef0f137deceb8fd8578cfadf0/conanfile.txt#L9. Then when the CMake project locates dependencies via calls to `find_package` then this can be overridden by Conan to point this to the local Conan dependencies if `conan install ...` is used to populate the build directory before configuring CMake (to do this Conan generates find modules within the build directory which override the default CMake find modules and point to the local Conan version of the libraries). If Conan is not used to install dependencies the CMake will search the standard system location by default when calling the standard CMake fine modules thus ensuring this works so long as these dependencies are found on the system.

**Reply by rui314** [reply]
mold is not a big project, so Conan might be too heavyweight for us. As to build dependencies, we have a simple shell script (https://github.com/rui314/mold/blob/main/install-build-deps.sh) which installs required packages.
